File: cses/05_range_queries/2206_pizzeria-queries/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    q = int(data[idx]); idx += 1

    INF = float('inf')
    N = n
    # Hai iterative segment tree với 2*N nút; lá nằm ở vị trí [N, 2N).
    # Tách trị tuyệt đối: L lưu p_a - a (dùng prefix-min), R lưu p_a + a (dùng suffix-min).
    L = [INF] * (2 * N)  # lưu p_a - a
    R = [INF] * (2 * N)  # lưu p_a + a

    # Nạp giá vào các lá theo chỉ số 1-indexed.
    for i in range(N):
        p = int(data[idx]); idx += 1
        a = i + 1  # 1-indexed
        L[N + i] = p - a
        R[N + i] = p + a

    # Dựng min ngược lên gốc cho cả hai cây.
    for i in range(N - 1, 0, -1):
        li = 2 * i
        L[i] = L[li] if L[li] < L[li + 1] else L[li + 1]
        R[i] = R[li] if R[li] < R[li + 1] else R[li + 1]

    out = []
    out_append = out.append

    for _ in range(q):
        t = data[idx]; idx += 1
        if t == b'2':
            # Truy vấn: chi phí nhỏ nhất khi đứng ở tòa k.
            k = int(data[idx]); idx += 1

            # prefix-min L trên [1, k] -> nửa khoảng 0-indexed [0, k), rồi cộng k.
            res = INF
            l = N
            r = k + N
            while l < r:
                if l & 1:
                    v = L[l]
                    if v < res:
                        res = v
                    l += 1
                if r & 1:
                    r -= 1
                    v = L[r]
                    if v < res:
                        res = v
                l >>= 1
                r >>= 1
            ans1 = res + k

            # suffix-min R trên [k, n] -> nửa khoảng 0-indexed [k-1, n), rồi trừ k.
            res = INF
            l = (k - 1) + N
            r = n + N
            while l < r:
                if l & 1:
                    v = R[l]
                    if v < res:
                        res = v
                    l += 1
                if r & 1:
                    r -= 1
                    v = R[r]
                    if v < res:
                        res = v
                l >>= 1
                r >>= 1
            ans2 = res - k

            # Đáp án là min của hai hướng trái/phải.
            out_append(ans1 if ans1 < ans2 else ans2)
        else:
            # Cập nhật: gán p_k = x, cập nhật cả hai cây rồi đẩy min lên gốc.
            k = int(data[idx]); idx += 1
            x = int(data[idx]); idx += 1
            i = k - 1

            # Cập nhật cây L với giá trị mới x - k.
            p = N + i
            L[p] = x - k
            p >>= 1
            while p:
                lp = 2 * p
                a = L[lp]; b = L[lp + 1]
                L[p] = a if a < b else b
                p >>= 1

            # Cập nhật cây R với giá trị mới x + k.
            p = N + i
            R[p] = x + k
            p >>= 1
            while p:
                lp = 2 * p
                a = R[lp]; b = R[lp + 1]
                R[p] = a if a < b else b
                p >>= 1

    sys.stdout.buffer.write(('\n'.join(map(str, out)) + ('\n' if out else '')).encode())


main()

File: cses/05_range_queries/2206_pizzeria-queries/solution.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <vector>

// Iterative segment tree (bottom-up) hỗ trợ range-min với point update.
class MinimumSegmentTree {
public:
    explicit MinimumSegmentTree(const std::vector<long long>& values)
        : size_(static_cast<int>(values.size())),
          tree_(2 * size_, infinity()) {
        // Gán giá trị vào các lá rồi dựng min ngược lên gốc.
        for (int index = 0; index < size_; ++index) {
            tree_[size_ + index] = values[index];
        }
        for (int index = size_ - 1; index > 0; --index) {
            tree_[index] = std::min(tree_[2 * index], tree_[2 * index + 1]);
        }
    }

    // Point update: gán giá trị mới cho một lá và cập nhật ngược lên gốc.
    void assign(const int position, const long long value) {
        int index = size_ + position;
        tree_[index] = value;
        for (index /= 2; index > 0; index /= 2) {
            tree_[index] = std::min(tree_[2 * index], tree_[2 * index + 1]);
        }
    }

    // Truy vấn min trên nửa khoảng [left, right).
    long long query(int left, int right) const {
        long long result = infinity();
        left += size_;
        right += size_;
        while (left < right) {
            if ((left & 1) != 0) {
                result = std::min(result, tree_[left++]);
            }
            if ((right & 1) != 0) {
                result = std::min(result, tree_[--right]);
            }
            left /= 2;
            right /= 2;
        }
        return result;
    }

private:
    static constexpr long long infinity() {
        return std::numeric_limits<long long>::max() / 4;
    }

    int size_;
    std::vector<long long> tree_;
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, q;
    std::cin >> n >> q;

    // Tách trị tuyệt đối: L[a] = p_a - a (cho a <= k), R[a] = p_a + a (cho a >= k).
    std::vector<long long> minus_index(n);
    std::vector<long long> plus_index(n);
    for (int position = 1; position <= n; ++position) {
        long long price;
        std::cin >> price;
        minus_index[position - 1] = price - position;
        plus_index[position - 1] = price + position;
    }

    // Cây trái giữ prefix-min của L, cây phải giữ suffix-min của R.
    MinimumSegmentTree left_tree(minus_index);
    MinimumSegmentTree right_tree(plus_index);

    while (q-- > 0) {
        int type, position;
        std::cin >> type >> position;
        if (type == 1) {
            // Cập nhật giá p_position = price ở cả hai cây.
            long long price;
            std::cin >> price;
            left_tree.assign(position - 1, price - position);
            right_tree.assign(position - 1, price + position);
        } else {
            // Đáp án = min( prefix-min L trên [1, k] + k , suffix-min R trên [k, n] - k ).
            const long long from_left =
                left_tree.query(0, position) + position;
            const long long from_right =
                right_tree.query(position - 1, n) - position;
            std::cout << std::min(from_left, from_right) << '\n';
        }
    }
    return 0;
}
